File: java/ReverseLinkedList.java

class Solution {

  /**
   * Function that reverses a linked list.
   * @param head the first node of the list.
   * @return the list in reverse order.
   */
  public ListNode reverseList(ListNode head) {
    ListNode previousNode = null;

    /**
     * While the head of the list is not null, iterate through it, and do the folllowing operations:
     *  1. set the node 'next' to be whatever is right after the head, so in the example it would be:
     *    head  = 1, next = 2, head.next = 2
     *  2. Set head.next to previous node, which at the time is currently null.
     *  3. The previous node gets the value of the head node.
     *  4. Head becomes next.
     */
    while (head != null) {
      ListNode next = head.next;
      head.next = previousNode;
      previousNode = head;
      head = next;
    }
    return previousNode;
  }
}

File: python/ReverseLinkedList.py
from typing import Optional

from python.ListNode import ListNode

class Solution:

    # Function that reverses a linked list.
    def reverse_list(self, head: Optional[ListNode]) -> Optional[ListNode]:

        # Create  a dummy previousNode, initially null. Point the currentNode to the head of the list. 
        previousNode = None
        currentNode = head

        while currentNode is not None:

            # Advance every pointer into its next node: next becomes the next of current. current.next = previous (no longer null), previous = current, current = next.
            nextNode = currentNode.next
            currentNode.next = previousNode
            previousNode = currentNode
            currentNode = nextNode

        # Return the last node of the list.
        return previousNode
